File: Linked List/206.Reverse-Linked-List.cpp
#include <bits/stdc++.h>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    void printLinkedList(ListNode* head) {
        while(head) {
            cout << head->val << " ";
            head = head->next;
        }
        cout << endl;
    }

    ListNode* reverseList(ListNode* head) {
        if (!head) {
            return head;
        }

        ListNode* curr = head;
        ListNode* prev = NULL;

        while(head) {
            curr = head->next;
            head->next = prev;
            prev = head;
            head = curr;
        }
        
        return prev;
    }


    ListNode* reverseListRecursion(ListNode* head) {
        if (!head) {
            return head;
        }

        ListNode* nextNode = head->next;
        if(!nextNode) {
            return head;
        }
        
        ListNode* newHead = reverseListRecursion(nextNode);
        nextNode->next = head;
        head->next = NULL;

        return newHead;
    }
};

int main() {
    ListNode* n1 = new ListNode(1);
    ListNode* n2 = new ListNode(2);
    ListNode* n3 = new ListNode(3);
    ListNode* n4 = new ListNode(4);
    n1->next = n2; n2->next = n3; n3->next = n4;
    
    Solution obj;
    obj.printLinkedList(n1);

    ListNode* listReversed = obj.reverseListRecursion(n1);
    obj.printLinkedList(listReversed);

    return 0;
}

File: LinkedList/206.Reverse-Linked-List.py
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev = None
        curr = head

        while curr:
            next_node = curr.next
            curr.next = prev
            prev = curr
            curr = next_node

        return prev
    
def create_linked_list(arr):
    if not arr:
        return None
    
    head = ListNode(arr[0])
    curr = head

    for val in arr[1:]:
        curr.next = ListNode(val)
        curr = curr.next 

    return head

def print_linked_list(head):
    curr = head

    while curr:
        print(curr.val, end="->")
        curr = curr.next

    print("None")

if __name__ == "__main__":
    arr = [1, 2, 3, 4, 5]
    head = create_linked_list(arr)
    print_linked_list(head)

    sol = Solution()
    print_linked_list(sol.reverseList(head))